api/envoy/type/metadata/v2/metadata.proto

```diff
@@ -0,0 +1,55 @@
+syntax = "proto3";
+
+package envoy.type.metadata.v2;
+
+option java_outer_classname = "MetadataProto";
+option java_multiple_files = true;
+option java_package = "io.envoyproxy.envoy.type.metadata.v2";
+
+import "validate/validate.proto";
+
+// [#protodoc-title: Metadata]
+
+// MetadataLocate provides a general interface using `filter` and `path` to retrieve value from
+// :ref:`Metadata <envoy_api_msg_core.Metadata>`.
+//
+// For example, for the following Metadata:
+//
+// .. code-block:: yaml
+//
+//    filter_metadata:
+//      envoy.filters.http.rbac:
+//        path:
+//          to:
+//            string: pro
+//
+// The following MetadataLocate is matched as the path [path, to, string]
+// will retrieve a string value "pro" from the Metadata.
+//
+// .. code-block:: yaml
+//
+//    filter: envoy.filters.http.rbac
+//    path:
+//    - key: path
+//    - key: to
+//    - key: string
+//
+message MetadataLocate {
+  // Specifies the segment in a path to retrieve value from Metadata.
+  // Note: Currently it's not supported to retrieve a value from a list in Metadata.
+  // This means that if the segment key refers to a list, it has to be the last segment in a path.
+  message PathSegment {
+    oneof segment {
+      option (validate.required) = true;
+
+      // If specified, use the key to retrieve the value in a Struct.
+      string key = 1 [(validate.rules).string = {min_bytes: 1}];
+    }
+  }
+
+  // The filter name to retrieve the Struct from the Metadata.
+  string filter = 1 [(validate.rules).string = {min_bytes: 1}];
+
+  // The path to retrieve the Value from the Struct.
+  repeated PathSegment path = 2 [(validate.rules).repeated = {min_items: 1}];
```

Comment:
It might be better to use the standard field mask (https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/field_mask.proto) or the field path (https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto#L125)

Reply:
copied from `MetadataMatcher`,  I will reconsider it again.

Reply:
one path in a field mask is flat field string joined with `'.'`, just like the original implementation in this PR (https://github.com/envoyproxy/envoy/blob/dc4ba7be17be8369c34a310b2a698c933a3878ac/api/envoy/type/tracing.proto#L70-L77)
and here we need only one path.
Also, seems that field_mask_util does not provide a native method to retrieve the value.

I wonder which one is better, the original flat type or structure type like that in `MetadataMatcher`. 
I am more and more inclined to still use the original model.

Reply:
WDYT, @mattklein123 @objectiser

Reply:
I like the simplicity of the original, but this current approach is consistent with the `MetadataMatcher` - so also not sure. Are there any usecases that could not be supported using the simpler approach? May be lists, although that is not currently supported by the `MetadataMatcher`, although possibly in the future?

Reply:
> May be lists, although that is not currently supported by the `MetadataMatcher`, although possibly in the future?

Yeah, i did also think about it. The type in `MetadataMatcher` provide a more convenient way to support it. Although the simpler approach can support some syntax like `[]`.

Reply:
Sorry what was the original? IMO it would be better to be consistent with the matcher if possible.

Reply:
Yeah, it's good. A common `MetadataKey` is provided there.